File: greatage-tapestry/src/main/resources/org/greatage/tapestry/commonlib/components/selector.js
/*
 * Copyright (c) 2008-2011 Ivan Khalopik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

Tapestry.SelectorItem = Class.create({
	initialize: function(elementId, value) {
		this._element = $(elementId);
		this._value = value;
	},

	value: function() {
		return this._value;
	},

	show: function() {
		this._element.removeClassName('invisible');
	},

	hide: function() {
		this._element.addClassName('invisible');
	}
});


Tapestry.Selector = Class.create({
	initialize: function(previousId, nextId, hiddenId, selectedId, items) {
		this._previous = $(previousId);
		this._next = $(nextId);
		this._hidden = $(nextId);

		this._current = 0;
		this._items = [];
		$H(items).each(function(entry, index) {
			this._items[index] = new Tapestry.SelectorItem(entry.key, entry.value);
			if (selectedId == entry.key) {
				this._current = index;
			}
		}, this);
		this.updateControls();
		this._previous.observe("click", this.onPrevious.bindAsEventListener(this));
		this._next.observe("click", this.onNext.bindAsEventListener(this));
	},

	previous : function() {
		if (this._current > 0) {
			this._items[this._current--].hide();
			this._items[this._current].show();
			this.updateControls();
		}
	},

	next : function() {
		if (this._current < this._items.length - 1) {
			this._items[this._current++].hide();
			this._items[this._current].show();
			this.updateControls();
		}
	},

	updateControls : function() {
		if (this._current > 0) {
			this._previous.removeClassName('disabled');
		} else {
			this._previous.addClassName('disabled');
		}
		if (this._current < this._items.length - 1) {
			this._next.removeClassName('disabled');
		} else {
			this._next.addClassName('disabled');
		}
		if (this._hidden) {
			this._hidden.value = this._items[this._current].value();
		}
	},

	onPrevious : function(event) {
		this.previous();
	},

	onNext : function(event) {
		this.next();
	}

});

Tapestry.Initializer.selector = function(spec) {
	new Tapestry.Selector(spec.previousId, spec.nextId, spec.hiddenId, spec.selectedId, spec.items);
};

File: greatage-tapestry/src/main/java/org/greatage/tapestry/commonlib/components/Selector.java
/*
 * Copyright (c) 2008-2011 Ivan Khalopik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.greatage.tapestry.commonlib.components;

import org.apache.tapestry5.BindingConstants;
import org.apache.tapestry5.ValueEncoder;
import org.apache.tapestry5.annotations.Environmental;
import org.apache.tapestry5.annotations.Import;
import org.apache.tapestry5.annotations.Parameter;
import org.apache.tapestry5.annotations.Property;
import org.apache.tapestry5.internal.TapestryInternalUtils;
import org.apache.tapestry5.ioc.annotations.Inject;
import org.apache.tapestry5.ioc.internal.util.CollectionFactory;
import org.apache.tapestry5.json.JSONObject;
import org.apache.tapestry5.services.ComponentDefaultProvider;
import org.apache.tapestry5.services.FormSupport;
import org.greatage.tapestry.CSSConstants;
import org.greatage.tapestry.commonlib.base.components.AbstractComponent;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author Ivan Khalopik
 * @since 1.0
 */
@Import(library = "selector.js")
public class Selector extends AbstractComponent {

	@Parameter
	private Iterable<?> source;

	@Parameter
	private Object value;

	@Parameter(name = "class", defaultPrefix = BindingConstants.LITERAL)
	private String className;

	@Property
	@Parameter
	private Object row;

	@Parameter(required = true)
	private ValueEncoder encoder;

	@Environmental(false)
	private FormSupport formSupport;

	@Inject
	private ComponentDefaultProvider defaultProvider;

	private Map<Object, String> itemIds;
	private Map<Object, String> itemValues;

	private String clientId;
	private String previousId;
	private String nextId;

	ValueEncoder defaultEncoder() {
		return defaultProvider.defaultValueEncoder("value", getResources());
	}

	public String getSelectorClass() {
		final List<String> classes = CollectionFactory.newList();
		classes.add(CSSConstants.CONTAINER_CLASS);
		classes.add(CSSConstants.SELECTOR_CLASS);
		if (className != null) {
			classes.add(className);
		}
		return TapestryInternalUtils.toClassAttributeValue(classes);
	}

	public boolean isVolatile() {
		return formSupport != null;
	}

	public String getClientId() {
		return clientId;
	}

	public String getPreviousId() {
		return previousId;
	}

	public String getPreviousClass() {
		return CSSConstants.SELECTOR_PREVIOUS_CLASS;
	}

	public String getNextId() {
		return nextId;
	}

	public String getNextClass() {
		return CSSConstants.SELECTOR_NEXT_CLASS;
	}

	public String getRowId() {
		return itemIds.get(row);
	}

	public String getRowClass() {
		return CSSConstants.SELECTOR_ITEM_CLASS + (row.equals(value) ? " " : " invisible");
	}

	void setupRender() {
		itemIds = new HashMap<Object, String>();
		itemValues = new HashMap<Object, String>();
		for (Object o : source) {
			final String itemId = allocateClientId("item");
			itemIds.put(o, itemId);
			itemValues.put(o, encoder.toClient(o));
		}
		if (value == null && source.iterator().hasNext()) {
			value = source.iterator().next();
		}

		clientId = allocateClientId();
		previousId = allocateClientId("previous");
		nextId = allocateClientId("next");
	}

	void afterRender() {
		final JSONObject spec = new JSONObject("previousId", previousId, "nextId", nextId, "hiddenId", clientId);
		spec.put("selectedId", itemIds.get(value));
		spec.put("items", generateItems());
		getJavaScriptSupport().addInitializerCall("selector", spec);
	}

	private JSONObject generateItems() {
		final JSONObject result = new JSONObject();
		for (Object item : source) {
			result.put(itemIds.get(item), itemValues.get(item));
		}
		return result;
	}
}
